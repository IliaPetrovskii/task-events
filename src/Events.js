/*
   1. Создайте функцию createButton(). Необходимо, чтобы эта функция осуществила вставку в body тег button с текстом: "Удали меня".
      При клике по button удалить этот button.
*/
export function createButton() {}

/*
   2. Создайте функцию createArrList(arr), в которую передается 1 параметр: arr - массив строк.
      Функция выводит этот массив в виде маркированного списка внутри тега body.
      При наведении курсора мыши на элемент списка у этого элемента создается атрибут title, в котором записан его текст.
*/
export function createArrList(arr) {}

/*
   3. Создайте функцию createLink(), которая сгенерирует следующую разметку и вставит ее в body:

      <a href="https://tensor.ru/">tensor</a>

      При первом клике по ссылке в конец ее текста через пробел дописывается ее href.
      При следующем клике происходит действие по умолчанию (переход по ссылке в текущей вкладке).
*/
export function createLink() {}

/*
   4. Создайте функцию createList(), которая сгенерирует следующую разметку и вставит ее в body:

      <ul>
         <li>Пункт</li>
      </ul>
      <button>Добавить пункт</button>

      При клике по элементу li ему в конец текста добавляется восклицательный знак.
      При клике по button в конец списка добавляется новый элемент li с текстом: "Пункт".
      Клик по новому li также добавляет восклицательный знак в конец текста.
*/
export function createList() {}
