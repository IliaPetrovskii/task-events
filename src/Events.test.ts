import { createButton, createArrList, createLink, createList } from './Events';

describe('События', () => {
    afterEach(() => {
        document.body.innerHTML = '';
    });

    describe('createButton()', () => {
        it('Создана кнопка в body с текстом "Удали меня"', () => {
            createButton();
            expect(document.body).toMatchSnapshot();
        });

        it('Кнопка удаляется по клику', () => {
            createButton();
            document.body.getElementsByTagName('button')[0].click();
            expect(document.body).toMatchSnapshot();
        });
    });

    describe('createArrList', () => {
        it('Создан список в body из элементов массива', () => {
            createArrList([
                'нулевой элемент',
                'первый элемент',
                'второй элемент',
            ]);
            expect(document.body).toMatchSnapshot();
        });

        it('По наведению на элементы списка создаётся атрибут title', () => {
            const arr = ['нулевой элемент', 'первый элемент', 'второй элемент'];
            createArrList(arr);

            function triggerEvent(el: HTMLElement): void {
                const e = new MouseEvent('mouseover', {
                    bubbles: true,
                    cancelable: true,
                });
                el.dispatchEvent(e);
            }

            for (let i = 0; i < arr.length; i++) {
                triggerEvent(document.body.getElementsByTagName('li')[i]);
            }

            expect(document.body).toMatchSnapshot();
        });
    });

    describe('createLink', () => {
        it('Создана ссылка "tensor" в body', () => {
            createLink();
            expect(document.body).toMatchSnapshot();
        });

        it('При первом клике в конец ссылки дописывается её href и блокируется действие по умолчанию', () => {
            createLink();

            const defaultHandlerStub = jest.fn();

            document.body
                .getElementsByTagName('a')[0]
                .addEventListener('click', defaultHandlerStub);

            document.body.getElementsByTagName('a')[0].click();
            expect(document.body).toMatchSnapshot();
            expect(defaultHandlerStub.mock.calls[0][0].defaultPrevented).toBe(
                true,
            );
        });

        it('При втором клике происходит действие по умолчанию', () => {
            createLink();

            const defaultHandlerStub = jest.fn();

            document.body
                .getElementsByTagName('a')[0]
                .addEventListener('click', defaultHandlerStub);

            document.body.getElementsByTagName('a')[0].click();
            document.body.getElementsByTagName('a')[0].click();

            expect(defaultHandlerStub).toBeCalledTimes(2);
            expect(defaultHandlerStub.mock.calls[1][0].defaultPrevented).toBe(
                false,
            );
            expect(
                defaultHandlerStub.mock.calls[1][0].target.getAttribute('href'),
            ).toBe('https://tensor.ru/');
        });
    });

    describe('createList', () => {
        it('Создан список и кнопка', () => {
            createList();
            expect(document.body).toMatchSnapshot();
        });

        it('При клике по элементу списка в конец текста добавляется восклицательный знак', () => {
            createList();
            document.body.getElementsByTagName('li')[0].click();
            expect(document.body).toMatchSnapshot();
        });

        it('При клике на кнопку добавляется новый элемент списка', () => {
            createList();
            document.body.getElementsByTagName('button')[0].click();
            expect(document.body).toMatchSnapshot();
        });

        it('При клике по новому элементу списка в конец текста добавляется восклицательный знак', () => {
            createList();
            document.body.getElementsByTagName('button')[0].click();
            document.body.getElementsByTagName('li')[1].click();
            expect(document.body).toMatchSnapshot();
        });
    });
});
