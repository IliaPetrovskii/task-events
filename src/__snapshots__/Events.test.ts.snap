// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`События createArrList По наведению на элементы списка создаётся атрибут title 1`] = `
<body>
  <ul>
    <li
      title="нулевой элемент"
    >
      нулевой элемент
    </li>
    <li
      title="первый элемент"
    >
      первый элемент
    </li>
    <li
      title="второй элемент"
    >
      второй элемент
    </li>
  </ul>
</body>
`;

exports[`События createArrList Создан список в body из элементов массива 1`] = `
<body>
  <ul>
    <li>
      нулевой элемент
    </li>
    <li>
      первый элемент
    </li>
    <li>
      второй элемент
    </li>
  </ul>
</body>
`;

exports[`События createButton() Кнопка удаляется по клику 1`] = `<body />`;

exports[`События createButton() Создана кнопка в body с текстом "Удали меня" 1`] = `
<body>
  <button>
    Удали меня
  </button>
</body>
`;

exports[`События createLink При первом клике в конец ссылки дописывается её href и блокируется действие по умолчанию 1`] = `
<body>
  <a
    href="https://tensor.ru/"
  >
    tensor https://tensor.ru/
  </a>
</body>
`;

exports[`События createLink Создана ссылка "tensor" в body 1`] = `
<body>
  <a
    href="https://tensor.ru/"
  >
    tensor
  </a>
</body>
`;

exports[`События createList При клике на кнопку добавляется новый элемент списка 1`] = `
<body>
  <ul>
    <li>
      Пункт
    </li>
    <li>
      Пункт
    </li>
  </ul>
  <button>
    Добавить пункт
  </button>
</body>
`;

exports[`События createList При клике по новому элементу списка в конец текста добавляется восклицательный знак 1`] = `
<body>
  <ul>
    <li>
      Пункт
    </li>
    <li>
      Пункт!
    </li>
  </ul>
  <button>
    Добавить пункт
  </button>
</body>
`;

exports[`События createList При клике по элементу списка в конец текста добавляется восклицательный знак 1`] = `
<body>
  <ul>
    <li>
      Пункт!
    </li>
  </ul>
  <button>
    Добавить пункт
  </button>
</body>
`;

exports[`События createList Создан список и кнопка 1`] = `
<body>
  <ul>
    <li>
      Пункт
    </li>
  </ul>
  <button>
    Добавить пункт
  </button>
</body>
`;
